File: src/BotIO.java
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.LinkedList;
    
public abstract class BotIO {

    private FileInputStream in;
    private FileOutputStream out;
    private boolean play = true;
    public Object bot;
    public int botNumber;

    public BotIO (int botNumber) {        
        try {
            this.botNumber = botNumber;
            bot = createBot(); 
            in = new FileInputStream("MatchToBot" + botNumber + ".txt");
            out = new FileOutputStream("BotToMatch" + botNumber + ".txt");
            play = true;
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void start() {
        try {
            while (play) {
                while (in.available() == 0){}
                String s = input();
                String[] lines = s.split("\n");
                for (int i = 0; i < lines.length; i++)
                    processInput(lines[i]);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void output(String s) {
        s = s + "\n";
        byte[] bytes = s.getBytes();
        try {
            out.write(bytes);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public String input() {
        try {
            LinkedList<Byte> bytes = new LinkedList<Byte>(); // This needs to be a list not an Array of size 
                                                             // in.available because a bot could theoretically
                                                             // print output while we are reading output
            for (byte b = 0; (b = (byte)(in.read())) != -1; )
                bytes.add(b);
            byte[] bytesArray = new byte[bytes.size()]; // can't use toArray because conversion from Byte to byte
            for (int i = 0; i < bytes.size(); i++)
                bytesArray[i] = (byte)bytes.get(i);
            return (new String(bytesArray)).trim();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    public void processInput(String s) {
        if (s == "")
            return; // TODO: process empty input error
        if (s.charAt(0) == 'X') {
            play = false;
            return;
        }
        output(sendToBot(s));
    }

    public abstract String sendToBot(String s);
    public abstract Object createBot();
} 

File: src/BotIO.py
import time

class BotIO:

    def __init__(self, botNum):
        self.botNum = botNum
        self.bot = self.createBot()
        self.infile = open("MatchToBot" + str(botNum) + ".txt",'r')
        self.outfile = open("Bot" + str(botNum) + "ToMatch.txt",'w')
        self.play = True    

    def start(self):
        while self.play:
           where = self.infile.tell()
           line = self.infile.readline()
           if not line:
               time.sleep(0.1)
               self.infile.seek(where)
           else:
               self.processInput(line)

    def output(self, s):
        if not len(s):
            return
        s += '\n'
        self.outfile.write(s)
        print('here')

    def processInput(self, s):
        if not len(s):
            return
        if s[0] == 'X':
            play = False
            return
        self.output(self.getBotResponse(s))

    def getBotResponse(self,s):
        raise NotImplementedError("Please implement this method.")

    def createBot(self):
        raise NotImplementedError("Please implement this method.")

